core/services/gateway/handlers/functions/orm.go

```diff
@@ -0,0 +1,136 @@
+package functions
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/lib/pq"
+	"github.com/pkg/errors"
+
+	"github.com/jmoiron/sqlx"
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/functions/generated/functions_router"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+)
+
+//go:generate mockery --quiet --name ORM --output ./mocks/ --case=underscore
+
+type ORM interface {
+	GetSubscriptions(offset, limit uint, qopts ...pg.QOpt) ([]CachedSubscription, error)
+	UpsertSubscription(subscription CachedSubscription, qopts ...pg.QOpt) error
+}
+
+type orm struct {
+	q                     pg.Q
+	routerContractAddress common.Address
+}
+
+var _ ORM = (*orm)(nil)
+var (
+	ErrInvalidParameters = errors.New("invalid parameters provided to create a subscription cache ORM")
+)
+
+const (
+	tableName = "functions_subscriptions"
+)
+
+type cachedSubscriptionRow struct {
+	SubscriptionID        uint64
+	Balance               string
+	Owner                 common.Address
+	BlockedBalance        string
+	ProposedOwner         common.Address
+	Consumers             pq.ByteaArray
+	Flags                 []uint8
+	RouterContractAddress common.Address
+}
+
+func NewORM(db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig, routerContractAddress common.Address) (ORM, error) {
+	if db == nil || cfg == nil || lggr == nil || len(routerContractAddress) == 0 {
+		return nil, ErrInvalidParameters
+	}
+
+	return &orm{
+		q:                     pg.NewQ(db, lggr, cfg),
+		routerContractAddress: routerContractAddress,
+	}, nil
+}
+
+func (o *orm) GetSubscriptions(offset, limit uint, qopts ...pg.QOpt) ([]CachedSubscription, error) {
+	var cacheSubscriptions []CachedSubscription
+	var cacheSubscriptionRows []cachedSubscriptionRow
+	stmt := fmt.Sprintf(`
+		SELECT subscription_id, owner, balance, blocked_balance, proposed_owner, consumers, flags, router_contract_address
+		FROM %s
+		WHERE router_contract_address = $1
+		ORDER BY subscription_id DESC
```

Comment:
DESC worries me a bit. It works now but it's different to how getSubscriptionsInRange() works and might cause some confusion in the future. Can we remove it?

Reply:
would you prefer to have ASC?
DESC is used just to have a defined order an be able to paginate with offset & limit.